src/runtime/virtcontainers/nydusd.go

```diff
@@ -0,0 +1,424 @@
+	Profile    string `json:"profile"`
+	Rustc      string `json:"rustc"`
+}
+
+type DaemonInfo struct {
+	ID      string        `json:"id"`
+	Version BuildTimeInfo `json:"version"`
+	State   string        `json:"state"`
+}
+
+type ErrorMessage struct {
+	Code    string `json:"code"`
+	Message string `json:"message"`
+}
+
+type MountRequest struct {
+	FsType string `json:"fs_type"`
+	Source string `json:"source"`
+	Config string `json:"config"`
+}
+
+func NewMountRequest(fsType, source, config string) *MountRequest {
+	return &MountRequest{
+		FsType: fsType,
+		Source: source,
+		Config: config,
+	}
+}
+
+type Interface interface {
+	CheckStatus() (DaemonInfo, error)
+	Mount(string, *MountRequest) error
+	Umount(sharedMountPoint string) error
+}
+
+type NydusClient struct {
+	httpClient *http.Client
+}
+
+func NewNydusClient(sock string) (Interface, error) {
+	transport, err := buildTransport(sock)
+	if err != nil {
+		return nil, err
+	}
+	return &NydusClient{
+		httpClient: &http.Client{
+			Timeout:   defaultHttpClientTimeout,
+			Transport: transport,
+		},
+	}, nil
+}
+
+func waitUntilSocketReady(sock string) error {
+	return retry.Do(func() error {
+		if _, err := os.Stat(sock); err != nil {
+			return err
+		}
+		return nil
+	},
+		retry.Attempts(3),
+		retry.LastErrorOnly(true),
+		retry.Delay(100*time.Millisecond))
+}
+
+func buildTransport(sock string) (http.RoundTripper, error) {
+	err := waitUntilSocketReady(sock)
+	if err != nil {
+		return nil, err
+	}
+	return &http.Transport{
+		MaxIdleConns:          maxIdleConns,
+		IdleConnTimeout:       idleConnTimeout,
+		ExpectContinueTimeout: expectContinueTimeout,
+		DialContext: func(ctx context.Context, _, _ string) (net.Conn, error) {
+			dialer := &net.Dialer{
+				Timeout:   dialTimout,
+				KeepAlive: keepAlive,
+			}
+			return dialer.DialContext(ctx, "unix", sock)
+		},
+	}, nil
+}
+
+func (c *NydusClient) CheckStatus() (DaemonInfo, error) {
+	resp, err := c.httpClient.Get(fmt.Sprintf("http://unix%s", infoEndpoint))
+	if err != nil {
+		return DaemonInfo{}, err
+	}
+	defer resp.Body.Close()
+	b, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return DaemonInfo{}, err
+	}
+	var info DaemonInfo
+	if err = json.Unmarshal(b, &info); err != nil {
+		return DaemonInfo{}, err
+	}
+	return info, nil
+}
+
+func (c *NydusClient) Mount(mountPoint string, mr *MountRequest) error {
+	requestURL := fmt.Sprintf("http://unix%s?mountpoint=%s", mountEndpoint, mountPoint)
+	body, err := json.Marshal(mr)
+	if err != nil {
+		return errors.Wrap(err, "failed to create mount request")
+	}
+
+	resp, err := c.httpClient.Post(requestURL, contentType, bytes.NewBuffer(body))
+	if err != nil {
+		return err
+	}
+	defer resp.Body.Close()
+	if resp.StatusCode == http.StatusNoContent {
+		return nil
+	}
+	return handleMountError(resp.Body)
+}
+
+func (c *NydusClient) Umount(sharedMountPoint string) error {
+	requestURL := fmt.Sprintf("http://unix%s?mountpoint=%s", mountEndpoint, sharedMountPoint)
```

Comment:
Same question as above: can this function assume `sharedMountPoint` is valid? I would suggest it shouldn't.